org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/InferenceContext18.java

```diff
@@ -1910,4 +1937,256 @@ public void cleanUp() {
 		this.b2 = null;
 		this.currentBounds = null;
 	}
+
+	// section 18.5.5
+	public ReferenceBinding inferRecordPatternParameterization(
+			RecordPattern recordPattern,
+			BlockScope scope2,
+			TypeBinding candidateT) {
+		TypeBinding typeBinding = recordPattern.resolvedType;
+		if (!(typeBinding instanceof ReferenceBinding))
+			return null; // should not happen.
+		// 1.If T is not downcast convertible (5.5) to the raw type R, inference fails.
+		Expression synthExpr = new Expression() {
+			@Override public StringBuffer printExpression(int indent, StringBuffer output) {
+				return output;
+			}
+		};
+		if (!synthExpr.checkCastTypesCompatibility(scope2, candidateT, typeBinding, synthExpr, false))
+			return null;
+//		if (!isDownwardCompatible(candidateT, typeBinding))
+//			return null;
+		//2. Otherwise, where P1, ..., Pn (n ≥ 1) are the type parameters of R,...
```

Comment:
I believe that within JLS 18.5.5 the check for downcast convertibility is the only thing that should actually use the raw type R (which we correctly get from SingleTypeReference.resolved()). This implies that after that check we can safely update typeBinding:
```typeBinding = typeBinding.original()``` which will allow us to remove all `original()` calls in downstream code for simplification.

PS: if you're comfortable with `synthExpr.checkCastTypesCompatibility()` you may want to remove the commented code above. :)

Reply:
Agree.. but I believe the //2.. needs to be kept for the steps.. keeping that, pl let me know if you think otherwise